src/hypofuzz/provider.py

```diff
@@ -26,29 +80,409 @@ def fresh_choice(
+            self._state.branches = frozenset(collector.branches)
+            # after_test_case is called by on_observation, since we want to wait
+            # until we know a few things about how the execution went
+
+    def before_test_case(self) -> None:
+        if not self._started:
+            self._startup()
+
+        assert self.corpus is not None
+        replay_priority = None
+        extra_queue_data = None
+        while self._replay_queue:
+            (replay_priority, choices, extra_queue_data) = self._replay_queue.pop()
+
+            # we checked if we had this choice sequence when we put it into the
+            # replay_queue on on_event, but we check again here, since we might have
+            # executed it in the interim.
+            if (
+                # Choices doesn't handle ChoiceTemplate
+                all(not isinstance(choice, ChoiceTemplate) for choice in choices)
+                and Choices(choices) in self.corpus.corpus
+            ):
+                continue
+
+            self._start_phase(Phase.REPLAY)
+            break
+        else:
+            # We stay in blackbox mode until we've generated 1000 consecutive examples
+            # without new coverage, and then switch to mutation.
+
+            # TODO: currently hard-coding a particular mutator; we want to do MOpt-style
+            # adaptive weighting of all the different mutators we could use.
+            # For now though, we'll just use a hardcoded swapover point
+            if self.since_new_branch < 1000 or self.random.random() < 0.05:
+                mutator: Mutator = BlackBoxMutator(self.corpus, self.random)
+            else:
+                mutator = CrossOverMutator(self.corpus, self.random)
+
+            choices = mutator.generate_choices()
+            self._start_phase(Phase.GENERATE)
+
+        start = time.perf_counter()
+        self.ninputs += 1
+
+        self._state = State(
+            choices=choices,
+            replay_priority=replay_priority,
+            start_time=start,
+            extra_queue_data=extra_queue_data,
+        )
+
+    def on_observation(
+        self, observation: Union[TestCaseObservation, InfoObservation]
+    ) -> None:
+        assert observation.type == "test_case"
+        self.after_test_case(observation)
+
+    def after_test_case(self, observation: TestCaseObservation) -> None:
+        assert self._state is not None
+        assert self._state.branches is not None
+        assert self.corpus is not None
+        assert self.database_key is not None
+        assert self.db is not None
+
+        assert observation.type == "test_case"
+        # run_start is relative to StateForActualGivenExecution, which we
+        # re-use per FuzzProcess. Overwrite with the current timestamp for use
+        # in sorting observations. This is not perfectly reliable in a
+        # distributed setting, but is good enough.
+        observation.run_start = time.time()
```

Comment:
let's make this "as of the start" by subtracting the increment we'll add to self.elapsed_time later.